flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetWriters.java

```diff
@@ -19,38 +19,436 @@
+    }
+  }
+
+  private static class FixedDecimalWriter extends ParquetValueWriters.PrimitiveWriter<DecimalData> {
+    private final int precision;
+    private final int scale;
+    private final ThreadLocal<byte[]> bytes;
+
+    private FixedDecimalWriter(ColumnDescriptor desc, int precision, int scale) {
+      super(desc);
+      this.precision = precision;
+      this.scale = scale;
+      this.bytes = ThreadLocal.withInitial(() -> new byte[TypeUtil.decimalRequiredBytes(precision)]);
+    }
+
+    @Override
+    public void write(int repetitionLevel, DecimalData decimal) {
+      byte[] binary = DecimalUtil.toReusedFixLengthBytes(precision, scale, decimal.toBigDecimal(), bytes.get());
+      column.writeBinary(repetitionLevel, Binary.fromReusedByteArray(binary));
+    }
+  }
+
+  private static class TimestampDataWriter extends ParquetValueWriters.PrimitiveWriter<TimestampData> {
+    private TimestampDataWriter(ColumnDescriptor desc) {
+      super(desc);
+    }
+
+    @Override
+    public void write(int repetitionLevel, TimestampData value) {
+      column.writeLong(repetitionLevel, value.getMillisecond() * 1000 + value.getNanoOfMillisecond() / 1000);
+    }
+  }
+
+  private static class ByteArrayWriter extends ParquetValueWriters.PrimitiveWriter<byte[]> {
+    private ByteArrayWriter(ColumnDescriptor desc) {
+      super(desc);
+    }
+
+    @Override
+    public void write(int repetitionLevel, byte[] bytes) {
+      column.writeBinary(repetitionLevel, Binary.fromReusedByteArray(bytes));
+    }
   }
 
-  public static ParquetValueWriter<Row> buildWriter(MessageType type) {
-    return INSTANCE.createWriter(type);
+  private static class ArrayDataWriter<E> extends ParquetValueWriters.RepeatedWriter<ArrayData, E> {
+    private final LogicalType elementType;
+
+    private ArrayDataWriter(int definitionLevel, int repetitionLevel,
+                            ParquetValueWriter<E> writer, LogicalType elementType) {
+      super(definitionLevel, repetitionLevel, writer);
+      this.elementType = elementType;
+    }
+
+    @Override
+    protected Iterator<E> elements(ArrayData list) {
+      return new ElementIterator<>(list);
+    }
+
+    private class ElementIterator<E> implements Iterator<E> {
+      private final int size;
+      private final ArrayData list;
+      private int index;
+
+      private ElementIterator(ArrayData list) {
+        this.list = list;
+        size = list.size();
+        index = 0;
+      }
+
+      @Override
+      public boolean hasNext() {
+        return index != size;
+      }
+
+      @Override
+      @SuppressWarnings("unchecked")
+      public E next() {
+        if (index >= size) {
+          throw new NoSuchElementException();
+        }
+
+        E element;
+        if (list.isNullAt(index)) {
+          element = null;
+        } else {
+          element = (E) ArrayData.createElementGetter(elementType).getElementOrNull(list, index);
```

Comment:
This method is called in a tight loop, so for performance any preparation that can be done in advance should be.

That means this getter should be created in the constructor and stored as an instance field. Then it can be called here.

Also, if there is already a null check above, does this need to call `getElementOrNull` or should it just call a `get` variant that assumes the value is non-null?

Alternatively, you could replace the `if` here:

```
E element = (E) getter.getElementOrNull(list, index);
```

Reply:
> That means this getter should be created in the constructor and stored as an instance field. Then it can be called here.

Yeah,  that sounds good to me, great point. 

> does this need to call getElementOrNull or should it just call a get variant that assumes the value is non-null?
The `getter` in ArrayData don't have a `get`  interface,  it have only the interface: 

```java
	/**
	 * Accessor for getting the elements of an array during runtime.
	 *
	 * @see #createElementGetter(LogicalType)
	 */
	interface ElementGetter extends Serializable {
		@Nullable Object getElementOrNull(ArrayData array, int pos);
	}
```

Replacing the `if-else` to be `E element = (E) getter.getElementOrNull(list, index);`  sounds reasonable to me.